### ISSUE TITLE
Poor performance on test data

### ISSUE THREAD

**Issue by saurabh502** [flairNLP/flair#333]
When I am training for text classification using FLAIR, I am getting excellent results, see below:
```
{'test_score': 0.9534,
 'dev_score_history': [0.9321,
  0.939,
  0.9449,
  0.9465,
  0.9456,
  0.9415,
  0.9512,
  0.9506,
  0.9543,
  0.9546], 
```
But when I am predicting on separate test file I am getting below par performance (F score below .15). Below is the code I am using for prediction:

```
def fun(x):
    sentence = Sentence(x)
    classifier.predict(sentence)
    return(sentence.labels)

test['label'] = test['text'].apply(fun)
test['label'] = test['text'].apply(lambda x : int(str(x[0]).split()[0]))
```

Any rationale behind this?

Thanks

**Comment by alanakbik** [comment]
Hi @saurabh502 that is indeed strange and should normally not happen. Maybe the two datasets are somehow differently preprocessed.

A few questions: 

- do the two datasets have different preprocessing, for instance different tokenization? 

- if you use this code on the test data split of the training data, do you get the same results as during training?

- could you try using `sentence = Sentence(x, use_tokenizer=True)` in the fun(x) function? this tokenizes input data.

- could you give us a sample of the training data and the separate test data?

**Comment by saurabh502** [comment]
Hi alanakbik,

Thank you for prompt response. Now that you have mentioned , I think preprocessing might be the case.
Below is the code I am using for train/valid/test split:

```
data['label'] = '__label__' + data['label'].astype(str)
data.iloc[0:int(len(data)*0.8)].to_csv('train.csv', sep='\t', index = False, header = False)
data.iloc[int(len(data)*0.8):int(len(data)*0.9)].to_csv('test.csv', sep='\t', index = False, header = False)
data.iloc[int(len(data)*0.9):].to_csv('dev.csv', sep='\t', index = False, header = False);
```

And below is the code for corpus:
`corpus = NLPTaskDataFetcher.load_classification_corpus(Path('./'), test_file='test.csv', dev_file='dev.csv', train_file='train.csv')`

Finally, below is the code for training the model:
```
word_embeddings = [WordEmbeddings('glove'), FlairEmbeddings('news-forward-fast'), FlairEmbeddings('news-backward-fast')]
document_embeddings = DocumentLSTMEmbeddings(word_embeddings, hidden_size=512, reproject_words=True, reproject_words_dimension=256)
classifier = TextClassifier(document_embeddings, label_dictionary=corpus.make_label_dictionary(), multi_label=False)
trainer = ModelTrainer(classifier, corpus)
trainer.train('./', max_epochs=2)
```

Results for train/test/valid are similar:

```
2018-12-27 19:17:32,211 ----------------------------------------------------------------------------------------------------
2018-12-27 19:17:32,212 Evaluation method: MICRO_F1_SCORE
2018-12-27 19:17:32,221 ----------------------------------------------------------------------------------------------------
2018-12-27 19:17:33,396 epoch 1 - iter 0/800 - loss 0.02252211
2018-12-27 19:18:09,279 epoch 1 - iter 80/800 - loss 0.00790335
2018-12-27 19:18:43,701 epoch 1 - iter 160/800 - loss 0.00792579
2018-12-27 19:19:15,877 epoch 1 - iter 240/800 - loss 0.00786001
2018-12-27 19:19:49,167 epoch 1 - iter 320/800 - loss 0.00774083
2018-12-27 19:20:22,466 epoch 1 - iter 400/800 - loss 0.00757523
2018-12-27 19:20:55,583 epoch 1 - iter 480/800 - loss 0.00751251
2018-12-27 19:21:30,613 epoch 1 - iter 560/800 - loss 0.00747602
2018-12-27 19:22:07,140 epoch 1 - iter 640/800 - loss 0.00744500
2018-12-27 19:22:44,750 epoch 1 - iter 720/800 - loss 0.00730330
2018-12-27 19:23:22,710 ----------------------------------------------------------------------------------------------------
2018-12-27 19:23:22,715 EPOCH 1 done: loss 0.0073 - lr 0.1000 - bad epochs 0
2018-12-27 19:24:10,364 DEV  : loss 0.00658339 - f-score 0.9277 - acc 0.9277
2018-12-27 19:25:01,054 TEST : loss 0.00676012 - f-score 0.9290 - acc 0.9290
2018-12-27 19:25:04,518 ----------------------------------------------------------------------------------------------------
2018-12-27 19:25:04,709 epoch 2 - iter 0/800 - loss 0.00704762
2018-12-27 19:25:18,076 epoch 2 - iter 80/800 - loss 0.00594975
2018-12-27 19:25:30,574 epoch 2 - iter 160/800 - loss 0.00588654
2018-12-27 19:25:43,358 epoch 2 - iter 240/800 - loss 0.00592454
2018-12-27 19:25:56,119 epoch 2 - iter 320/800 - loss 0.00600569
2018-12-27 19:26:08,945 epoch 2 - iter 400/800 - loss 0.00614073
2018-12-27 19:26:22,099 epoch 2 - iter 480/800 - loss 0.00618101
2018-12-27 19:26:35,560 epoch 2 - iter 560/800 - loss 0.00614024
2018-12-27 19:26:47,717 epoch 2 - iter 640/800 - loss 0.00616961
2018-12-27 19:27:00,498 epoch 2 - iter 720/800 - loss 0.00611484
2018-12-27 19:27:12,629 ----------------------------------------------------------------------------------------------------
2018-12-27 19:27:12,634 EPOCH 2 done: loss 0.0061 - lr 0.1000 - bad epochs 0
2018-12-27 19:27:25,494 DEV  : loss 0.00548878 - f-score 0.9334 - acc 0.9334
2018-12-27 19:27:39,234 TEST : loss 0.00570648 - f-score 0.9334 - acc 0.9334
2018-12-27 19:27:45,798 ----------------------------------------------------------------------------------------------------
2018-12-27 19:27:45,805 Testing using best model ...
2018-12-27 19:28:00,063 MICRO_AVG: acc 0.9334 - f1-score 0.9334
2018-12-27 19:28:00,068 MARCO_AVG: acc 0.565 - f1-score 0.871
2018-12-27 19:28:00,076 0          tp: 2962 - fp: 208 - fn: 5 - tn: 2962 - precision: 0.9344 - recall: 0.9983 - accuracy: 0.9653 - f1-score: 0.9653
2018-12-27 19:28:00,084 1          tp: 21 - fp: 5 - fn: 208 - tn: 21 - precision: 0.8077 - recall: 0.0917 - accuracy: 0.1647 - f1-score: 0.1647
2018-12-27 19:28:00,089 ----------------------------------------------------------------------------------------------------
{'test_score': 0.9334,
 'dev_score_history': [0.9277, 0.9334],
 'train_loss_history': [0.007278426149265222, 0.00607920798072288],
 'dev_loss_history': [0.006583388429135084, 0.00548877939581871]}
```
But when I am predicting on separate data results are not as expected, I am using below code for prediction:
```
classifier = TextClassifier.load_from_file('./best-model.pt')
def fun(x):
    sentence = Sentence(x, use_tokenizer=True)
    classifier.predict(sentence)
    return(sentence.labels)
test['label'] = test['text'].apply(fun)

```

So the question here is how should I preprocess data before predicting?

Thanks

**Comment by alanakbik** [comment]
Hi - this part of code looks good. In fact, since it looks like you're not using the tokenization option during training, you can probably go back to without the use_tokenizer=True option. 

Maybe the problem is in the `test['label'] = test['text'].apply(fun)` line. Here, `test` is only one instance you classify, right? Could you also paste the rest of the code, i.e. how do you iterate through the test dataset and write out the predictions, and how do you then evaluate? Maybe there is a problem in how you write out predictions before evaluation.

Also, could you manually check the prediction in `test['label']` with some examples that should work? Is the prediction good or not good?

**Comment by alanakbik** [comment]
Hi, you can get the scores by accessing the relevant fields ('value' and 'score') in the `Label` object, like this: 

```python
# example classifier
text_classifier = TextClassifier.load('en-sentiment')

# example sentence
sentence = Sentence('This is great!')

# predict labels for sentence
text_classifier.predict(sentence)

# iterate through predicted labels
for label in sentence.labels:
    print(label)
    print(label.value)
    print(label.score)
```

This should print: 

```console
POSITIVE (0.5852776765823364)
POSITIVE
0.5852776765823364
```

Wrt to the original problem: Could you try manually checking the predictions for some texts that should work? Does the classifier make correct predictions?

**Comment by saurabh502** [comment]
Hi alanakbik,

Thank you for your help in this. To further dig into the issue , i did following:
1) I created a copy of test.csv file and renamed it test1.csv from below steps:

```
data['label'] = '__label__' + data['label'].astype(str)
data.iloc[0:int(len(data)*0.8)].to_csv('train.csv', sep='\t', index = False, header = False)
data.iloc[int(len(data)*0.8):int(len(data)*0.9)].to_csv('test.csv', sep='\t', index = False, header = False)
data.iloc[int(len(data)*0.9):].to_csv('dev.csv', sep='\t', index = False, header = False);
```
`data.iloc[int(len(data)*0.8):int(len(data)*0.9)].to_csv('test1.csv', index = False)`

2) Trained the classifier 

```
corpus = NLPTaskDataFetcher.load_classification_corpus(Path('./'), test_file='test.csv', dev_file='dev.csv', train_file='train.csv')

word_embeddings = [WordEmbeddings('glove'), FlairEmbeddings('news-forward-fast'), FlairEmbeddings('news-backward-fast')]
document_embeddings = DocumentLSTMEmbeddings(word_embeddings, hidden_size=512, reproject_words=True, reproject_words_dimension=256)
classifier = TextClassifier(document_embeddings, label_dictionary=corpus.make_label_dictionary(), multi_label=False)
trainer = ModelTrainer(classifier, corpus)
trainer.train('./', max_epochs=2)
```
3) Finally predicted on the test1.csv

```
classifier = TextClassifier.load_from_file('./best-model.pt')

def fun(x):
    sentence = Sentence(x)
    classifier.predict(sentence)
    for label in sentence.labels:        
        return(label.value)
        
%%time
test['label1'] = test['tweet'].apply(fun)
```

4) But when I compared results with actual target, it was poor
 
**predicted label**
```
0    3180
1      16
Name: label1, dtype: int64
```
**actual label**

```
0    2959
1     237
Name: label, dtype: int64
```

```
              precision    recall  f1-score   support

           0       0.93      1.00      0.96      2959
           1       0.75      0.05      0.09       237

   micro avg       0.93      0.93      0.93      3196
   macro avg       0.84      0.52      0.53      3196
weighted avg       0.92      0.93      0.90      3196
```
But in model output the result for test is excellent:
```
{'test_score': 0.9334,
 'dev_score_history': [0.9277, 0.9334],
 'train_loss_history': [0.007278426149265222, 0.00607920798072288],
 'dev_loss_history': [0.006583388429135084, 0.00548877939581871]}
```

Is this because the data is unbalanced and the model is reporting micro avg of f1-score?
One connected question , how can we get other evaluation metrics in model output in order to evaluate model better?

Thanks in advance for your help in this!

**Comment by alanakbik** [comment]
Hi @saurabh502 thanks for looking into this! 

Yes, it could be because of unbalanced data and the micro average. If you want to switch evaluation metrics during training, you can do this by passing the `evaluation_metric` parameter to the trainer, like this: 

```python
trainer.train('./',
              evaluation_metric=EvaluationMetric.MACRO_F1_SCORE,
              max_epochs=2) 
```

But no matter what you pass, the trainer always reports all metrics after the last epoch of training, i.e. both micro and macro F1 and accuracies. 

For better results, also consider increasing the `max_epochs` to a higher value to give the trainer for time to learn.

**Comment by tsu3010** [comment]
> Hi Alanakbik,
> 
> test is a csv file which i have converted to pandas dataframe, below is sample of test:
> 
> ```
> 	id	        text
> 0	31963	#studiolife #aislife #requires #passion #dedic...
> 1	31964	@user #white #supremacists want everyone to s...
> 2	31965	safe ways to heal your #acne!! #altwaystohe...
> 3	31966	is the hp and the cursed child book up for res...
> 4	31967	3rd #bihday to my amazing, hilarious #nephew...
> ```
> 
> So using apply I am predicting label for each text columns, which i am processing using below code:
> `test['label'] = test['label'].apply(lambda x : int(str(x[0]).split()[0]))`
> 
> Because predict gives output in below format:
> [0 (1.0)]
> 
> Hope this clarifies my issue.
> 
> On a separate note, how can we get probabilities instead hard classes 0/1.
> 
> Thanks


Hi @alanakbik , 

I've noticed this in the test.tsv generated as well. Although f1 score is reasonable(at around 0.8 for multi class classification), all probability outputs are 1.0. Using the predict function like you mention below to obtain probability! 

```
clf.predict(sentence)
for label in sentence.labels:
    print(label)
    print(label.value)
    print(label.score)
```

But the label.score always returns a 1.0! Using v0.4.1 of flair currently. Any thoughts on why this could be happening?

**Comment by shengchien** [comment]
I noticed the discrepancy at first during my model training. After tracing the source code, I found it's because of different default tokenizer being used.

Both NLPTaskDataFetcher.load_classification_corpus and CSVClassificationDataset class (which is the one I use) default to "segtok_tokenizer".

https://github.com/zalandoresearch/flair/blob/v0.4.4/flair/data_fetcher.py#L399
https://github.com/zalandoresearch/flair/blob/v0.4.4/flair/datasets.py#L927

However, Sentence class defaults to "space_tokenizer"

https://github.com/zalandoresearch/flair/blob/v0.4.4/flair/data.py#L565

By replacing Sentence(x) with Sentence(text, use_tokenizer=True) like @alanakbik suggested gave me the same excellent result reported by ModelTrainer. Hope this helps.

**Comment by gokulprasadthekkel** [comment]
Hi, i'm having the same issue. Setting use_tokenize=True didn't help. I'm wondering whether the classification report generated by flair has some bug
